optuna/storages/_grpc/servicer.py

```diff
@@ -329,11 +347,15 @@ def GetTrials(
         study_id = request.study_id
         included_trial_ids = set(request.included_trial_ids)
         trial_id_greater_than = request.trial_id_greater_than
-        try:
-            trials = self._backend.get_all_trials(study_id, deepcopy=False)
-        except KeyError as e:
-            context.abort(code=grpc.StatusCode.NOT_FOUND, details=str(e))
-
+        if self._ttl_cache_seconds is None:
+            trials = self._get_all_trials(study_id, context)
+        else:
+            now = time.time()
+            if now >= self._cached_trials[study_id].last_sync_time + self._ttl_cache_seconds:
+                _logger.info("Syncing trials for study_id=%d", study_id)
+                self._cached_trials[study_id].trials = self._get_all_trials(study_id, context)
+                self._cached_trials[study_id].last_sync_time = now
+            trials = self._cached_trials[study_id].trials
```

Comment:
In real-world use cases, it's possible for multiple Kubernetes pods to send `GetAllTrials` requests simultaneously. Since `self._get_all_trials(study_id, context)` is a time-consuming operation, the cache becomes ineffective in such situations. For example, if 10 requests arrive at the same time, `self._get_all_trials(study_id, context)` will be called 10 times.

To prevent this, you could wrap the relevant section with a `threading.Lock` to ensure that only one thread performs the `self._get_all_trials` call while others wait for the result to be cached.

Reply:
thanks, I fixed it.